### PR TITLE
Ignore DNSKEY RRs with incalculable key sizes

### LINKED ISSUE #110: DNSSEC/dnssec03 and DNSSEC/dnssec05 TC crash in case of "empty" keys
Here is a domain with zero length RSA key
```
nlagriculture.nl.	86400 IN DNSKEY	257 3 13 (
				vRMOgGXuo/RaeD1XNWmDPvSDQDCNRa68OTq+/BqucA25
				ACppkunDt5fnRF2IL76ZpybyG7W8IBNYj7dpYrzWOg==
				) ; KSK; alg = ECDSAP256SHA256 ; key id = 12993
nlagriculture.nl.	86400 IN DNSKEY	256 3 8 (
				AwEAAeBjJKDZyqH6QoTM3lSHZ2Gk+26y6S/MH1Va2noA
				MfGAu1t2jg8Ve1m1dR53hNjaQpDiKkwg9UNH7wCdEKPT
				/PoAMP3Q3UZnqnoPk4b2wwyGEN5v+eIvDREoc6Eg1leA
				A8RQ2/GatjqWVR5K7wQ1UFCSyIOhgiPYpOKqfoFAnmx1
				) ; ZSK; alg = RSASHA256 ; key id = 18111
nlagriculture.nl.	86400 IN DNSKEY	257 3 8 (
				
				) ; KSK; alg = RSASHA256 ; key id = 1033
nlagriculture.nl.	86400 IN DNSKEY	256 3 8 (
				AwEAAfc58Rv7aL6+1cVlGaGF6P5mjxE77hDFXAKJFukO
				prucKqO1zupdHGDHrkWlvnBYxR+MArXPqZxsXAjNNgdb
				HkCcKG4S+qho+o1hjalGiTmslsdKCTlFoSf7u6Q4rlVf
				r/yUbF09cQdCMdjBEjcnezRBSAUNa7zu6fPPDdZJ/tfj
				) ; ZSK; alg = RSASHA256 ; key id = 61578
```
both test cases crash with the following message

```
   2.62 INFO      The DNSKEY with tag 61578 uses algorithm number 8 (RSA/SHA-256), which is OK.
   2.62 DEBUG     Key with keytag 61578 details : Size = 1024, Flags (SEP bit *not* set, RFC 5011 revocation bit *not* set).
   2.62 INFO      The DNSKEY with tag 18111 uses algorithm number 8 (RSA/SHA-256), which is OK.
   2.62 DEBUG     Key with keytag 18111 details : Size = 1024, Flags (SEP bit *not* set, RFC 5011 revocation bit *not* set).
   2.62 INFO      The DNSKEY with tag 1033 uses algorithm number 8 (RSA/SHA-256), which is OK.
perl: ./rdata.c:40: ldns_rdf_data: Assertion `rd != NULL' failed.
```

### PR DESCRIPTION
**Pull request by mattias-p**
## Purpose

This PR makes us ignore DNSKEY RRs with empty public key fields instead of crashing.

## Context

Fixes #110.
Implements the specification updates from zonemaster/zonemaster#1041.

## Changes

* Zonemaster::LDNS::RR::DNSKEY::keydata() is updated to work with ldns's internal representation of an empty DNSKEY public key field. (An issue has been created about changing the internal interpretation: nlnetlabs/ldns#156)
* Zonemaster::LDNS::RR::keysize() is ported from XS to Perl.
* Zonemaster::LDNS::RR::keysize() is fixed to return the `-1` sentinel value instead of crashing or returning a negative number of bits.
* Zonemaster::LDNS::Package::answer() is changed to filter out DNSKEY RRs with incalculable key sizes. The old behavior is still available under a new name: Zonemaster::LDNS::Package::answer_unfiltered().

## How to test this PR

A test case is provided in #110.

### Notes

After this PR when a zone with a single bad DNSKEY whose key size is incalculable, we get the following behaviors:
* DNSSEC03 emits a NO_DNSKEY message.
* DNSSEC07 emits a DS_BUT_NOT_DNSKEY message.
* DNSSEC05 emits one NO_RESPONSE_DNSKEY message for each name server that returned the bad DNSKEY record.
* DNSSEC14 emits one NO_RESPONSE_DNSKEY message for each name server that returned the bad DNSKEY record.

In principle I believe these are reasonable behaviors, but at least some of the tag names are a bit misleading. Especially the word "response" is unfortunate in this context. Perhaps it would also make sense to qualify "dnskey" with the adjective "valid" since there might actually be an RR in the answer section with the DNSKEY type that just happens to be broken enough that it is ignored.

### CODE REVIEW

**Comment by mattias-p** [comment]
I rebased onto the "Split section accessors into XS and Perl parts" commit from #136 in order to make the conflict resolution simpler when merging the second one.

**Comment by matsduf** [comment]
How does this PR relate to #136? I see similar or identical changes in both.

File: t/rr.t [hunk 1]
```diff
@@ -122,9 +136,9 @@ subtest 'RRSIG' => sub {
             is( $rr->signer, 'se.' );
             is( $rr->labels, 1 );
             if ( $rr->typecovered eq 'DNSKEY' ) {
-                # .SE KSK should not change very often. 59407 will replace 59747.
-                # Now (December 2017) both are used.
-                ok( $rr->keytag == 59747 or $rr->keytag == 59407 );
+                # .SE KSK should not change very often. 59407 has replaced 59747.
+                # Now (February 2022) only 59407 is used.
+                ok( $rr->keytag == 59407 );
```
Comment (matsduf) [inline t/rr.t:141]: This is not related to this PR as defined in the description. Move this to a separate PR. More of that below.

**Review by pnax — APPROVED** [review]
Nice! Tanks for the detailed description and documentation. I agree it would be nice to have better messages. In the meantime this is good to me.

**Review by tgreenx — APPROVED** [review]
LGTM, a welcome change

**Comment by mvw-afnic** [comment]
That’s odd: according to my testing, the unit tests pass, but by following the testing procedure in this PR, I still get crashes.

I saved `data-good.txt` and `data-bad.txt` from [here](https://github.com/zonemaster/zonemaster-ldns/issues/110#issuecomment-1061492309) in a directory called “data”, merged the PR locally, then built a Docker image for zonemaster-cli. Then I ran the following command:

```
$ docker run --rm --volume=$PWD/data:/data \
       zonemaster/cli:local \
       --no-ipv6 --test dnssec --restore /data/data-bad.txt nlagriculture.nl
Seconds Level     Message
======= ========= =======
Assertion failed: rd != NULL (./rdata.c: ldns_rdf_size: 26)
```

Yet I could verify that my Docker image had indeed the modified version of Zonemaster::LDNS.

What’s even stranger, is that running the same test against the current state of `develop` gives this result:
```
Seconds Level     Message                                                                  
======= ========= =======                                                                  
   0.67 ERROR     The DNSKEY record with tag 12993 that the DS refers to does not exist in the DNSKEY RRset. Fetched from the nameservers with IP "145.100.177.67;178.22.85.27;185.136.96.82;94.228.142.136".                                                                    
Assertion failed: rd != NULL (./rdata.c: ldns_rdf_size: 26)
```

Maybe `data-{good,bad}.txt` are wrong in some way, or have broken DNSKEYs in ways that this pull request hasn’t yet guarded against?

**Comment by mvw-afnic** [comment]
It seems that `data-bad.txt` broke DNSKEY RRs differently than what issue #110 described.

I found one instance of a DNSKEY record that is completely truncated (i.e. with zero RDLENGTH) instead of only missing its public key. Maybe this is why my testing didn’t work. Am I right?
